audio_helper/src/openxtalsa.c

```diff
@@ -0,0 +1,955 @@
+
+            // Check for EPIPE (xrun / suspended). If this is the case, we
+            // restart ALSA and try again.
+            if (ret == -EPIPE) {
+                ret = openxt_alsa_prepare(settings);
+                openxt_assert_ret(ret == 0, ret, ret);
+                continue;
+            }
+
+            // If the error is EAGAIN, it means that we are in non-blocking
+            // mode, and that we should try again later. In this case, we
+            // report 0 bytes.
+            if (ret == -EAGAIN) {
+                ret = 0;
+                break;
+            }
+
+            // If we got this far, we got have an error.
+            openxt_error("snd_pcm_readi failed: %d - %s\n", ret, snd_strerror(ret));
+            break;
+        }
+
+        // Done
+        break;
+    }
+
+    // Done
+    return ret;
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Simple Element Functions                                                                            //
+/////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+int openxt_alsa_mixer_fini(Settings *settings)
+{
+    int ret = 0;
+
+    // Sanity checks
+    openxt_checkp(settings, -EINVAL);
+
+    // Cleanup
+    if (settings->mhandle != NULL){
+
+        // Cleanup
+        snd_mixer_free(settings->mhandle);
+
+        // Close the mixer handle
+        ret = snd_mixer_close(settings->mhandle);
+    }
+
+    // Reset
+    settings->mhandle = NULL;
+
+    // Done
+    return ret;
+
+}
+
+int openxt_alsa_mixer_init(Settings *settings)
+{
+    int ret;
+
+    // Sanity checks
+    openxt_checkp(settings, -EINVAL);
+    openxt_assert_quiet(settings->mhandle == NULL, 0);
+
+    // Setup the handle
+    ret = snd_mixer_open(&settings->mhandle, 0);
+    openxt_assert_goto(ret == 0, falure);
+    ret = snd_mixer_attach(settings->mhandle, device);
+    openxt_assert_goto(ret == 0, falure);
+    ret = snd_mixer_selem_register(settings->mhandle, NULL, NULL);
+    openxt_assert_goto(ret == 0, falure);
+    ret = snd_mixer_load(settings->mhandle);
+    openxt_assert_goto(ret == 0, falure);
+
+    // Success
+    return 0;
+
+falure:
+
+    // Safety
+    openxt_alsa_mixer_fini(settings);
+
+    // Failure
+    return ret;
+}
+
+int openxt_alsa_mixer_print_selement(Settings *settings)
+{
+    int ret = 0;
+    int btn = 0;
+    long vol = 0;
+    long min = 0;
+    long max = 0;
+    char type = 'D';
+    bool has_volume = false;
+    bool has_switch = false;
+
+    char name_string[256];
+    char volume_string[256];
+    char switch_string[256];
+
+    snd_mixer_selem_id_t *sid = NULL;
+
+    // Sanity checks
+    openxt_checkp(settings, -EINVAL);
+    openxt_checkp(settings->elem, -EINVAL);
+    openxt_checkp(settings->mhandle, -EINVAL);
+
+    // Clear the strings for safety.
+    memset(name_string, 0, sizeof(name_string));
```

Comment:
I like being on the safe side, but in the interest of time..... I removed them :)